libvast/src/system/explorer.cpp

```diff
@@ -122,16 +146,37 @@ explorer(caf::stateful_actor<explorer_state>* self, caf::actor node,
         if (st.after)
           after_expr = predicate{attribute_extractor{timestamp_atom::value},
                                  less_equal, data{*x + *st.after}};
-        expression expr;
-        if (after_expr && before_expr) {
-          expr = conjunction{*before_expr, *after_expr};
-        } else if (after_expr) {
-          expr = *after_expr;
-        } else if (before_expr) {
-          expr = *before_expr;
+        std::optional<vast::expression> by_expr;
+        if (st.by) {
+          VAST_ASSERT(by_column); // Should have been checked above.
+          auto ci = (*by_column)[i];
+          if (caf::get_if<caf::none_t>(&ci))
+            continue;
+          // TODO: Make `predicate` accept a data_view as well to save
+          // the call to `materialize()`.
+          by_expr = predicate{key_extractor{*st.by}, equal, materialize(ci)};
         }
-        VAST_TRACE(self, "constructed expression", expr);
-        auto query = to_string(expr);
+        auto build_conjunction
```

Comment:
```suggestion
        auto make_conjunction
```

Pretty much all factory functions have a `make_` prefix in the code base. Using this here as well would improve consistency.

Reply:
Speaking of this, I think this whole helper function could quite easily be made variadic. It's three conjunctions now and it'll only become more in the future.

Reply:
The principle of using the least complex mechanism to achieve a goal would guide me to keep this as is. If we need it more flexible later on, then we can refactor still.

Reply:
This works and I don't think it's any harder to read nor is it a lot more complex.

```cpp

template <class Lhs, class... Rhs>
auto make_conjunction(Lhs&& lhs, Rhs&&... rhs) -> std::optional<expression> {
  if constexpr (sizeof...(rhs) == 0)
    return std::forward<Lhs>(lhs);
  else {
    auto rhs_eval = make_conjunction(std::forward<Rhs>(rhs)...);
    if (lhs && rhs_eval)
      return conjunction{*std::forward<Lhs>(lhs), std::move(*rhs_eval)};
    if (lhs)
      return std::forward<Lhs>(lhs);
    return std::move(rhs_eval);
  }
};
```

```cpp
        auto expr = make_conjunction(std::move(before_expr),
            std::move(after_expr), std::move(by_expr));
```

Reply:
But this is a big difference! Now the function is recursive and cluttered with `move` and `forward`. The metal toll for the reader/reviewer has significantly increased.

Even though it's pretty, we don't have a use case for it *today*. When we need such heavy machinery, we can still deploy this. Until then, I would prefer to prioritize KISS.

Reply:
I'd argue the readability on the call site improves a lot and is arguably more important to the reader. Anyways, we can revisit this discussion once we have more than three conjunctions at the call site.

This whole verbosity with `optional` could've been avoided with an always-true expression, which is something that we should maybe consider adding. `T` gets expanded to `:bool == T`, so that can't just work. It's just a useful thing to have in general for crafting queries.

Reply:
At the risk of beating the dead horse, but this is also not yet at its final stage of complexity because we'd still need to add a recursive static_assert that all types are `std::optional<expression>`.

Reply:
You'd just need to check that for Lhs, which is simple enough. 😅